### PR TITLE
Improve with AsyncState type definition

### PR DESCRIPTION
**Pull request by Rogdham**
## Issue

The `AsyncState` type could return an object with `loading` being false, but no `value` nor `error`.

This makes handling the `value` returned by e.g. `useAsync` quite difficult:

code

<details><summary>Code used for easy copy/paste</summary>
<p>

```ts
import { useAsync } from 'react-use';

const getPromise = async (): Promise<number> => 6;

const test = (): string => {
  const state = useAsync(getPromise, []);
  if (state.loading === true) {
    return 'loading...';
  }
  if (state.error !== undefined) {
    return state.error.message;
  }

  const result = 7 * state.value;
  return `${result}`;
};

export default test;
```

</p>
</details>

In the above example, the loading and error cases have already been ruled out, so `value` should not be `undefined`, creating issues:
 * This makes developers test if the `value` is undefined or not, which is an useless test;
 * If the `Promise` was actually returning a value that could be undefined, then we have no assurance that the (undefined) value actually comes from the promise.

### Commit creating the issue

In https://github.com/streamich/react-use/pull/348, @NullVoxPopuli merged the type `AsyncState` in `useAsync` with the one in `useAsyncFn`. However, there were different!

```diff
*** useAsync_type.ts
--- useAsyncFn_type.ts
***************
*** 1,7 ****
  export type AsyncState<T> =
    | {
!       loading: true;
        error?: undefined;
        value?: undefined;
      }
    | {
--- 1,7 ----
  export type AsyncState<T> =
    | {
!       loading: boolean;
        error?: undefined;
        value?: undefined;
      }
    | {
```

The fact that `loading` was `true` and not `boolean` solved the issue in `useAsync`, but it is not possible in `useAsyncFn` because of the initial value.

### Possibly related issue

@wardoost 's screenshot in https://github.com/streamich/react-use/issues/190#issuecomment-478305429 seems related even though the issue may not be

## Proposed solution

In this PR, I revert the diff above in the type `AsyncState`, and overload the type of `useAsyncFn` to  make the best of both worlds.

In other words, for the type `AsyncState<T>` if `loading` is `false`, then either `error` (of type `Error`) or `value` (of type `T`) has to be specified.

code2

### Breaking change

Before this PR, it was possible to call `useAsyncFn` with an `initalState` set to `{ loading: false }`, although probably nobody did because it was the default if no `initialState` was provided.

To replicate the behaviour, you can either:
 * use `{ loading: false, value: undefined }` as the `initialState`
 * do not specify `initialState` at all, since it is the default value

I would argue that this breaking change is minor as people in that case would have probably chosen to not specify `initialState` already.

### CODE REVIEW

**Comment by streamich** [comment]
In your example, it seems you can replace this line

```ts
const result = 7 * state.value;
```

by this (not `!` non-null assertion):

```ts
const result = 7 * state.value!;
```

would that solve your case?

**Comment by Rogdham** [comment]
> would that solve your case?

Thank you for your reply! That solves the typing issue (an other idea would have been testing if `value` is `undefined`), but does not solve the issue at hand. I would argue that it's more like a workaround against a weak typing of the returned value.

See, as a TypeScript user of `useAsync`, I don't want to look at the way `useAsync` is implemented, I should be able to rely exclusively on docs and types.

 * On one hand, reading [the doc](https://github.com/streamich/react-use/blob/master/docs/useAsync.md), I would have assumed that if `loading` is `false` and `error` is `undefined`, then the `value` is returned directly.
 * On the other hand, looking at the returned types let me think that there is a case in which `useAsync` returns an object `{ loading: false, error: undefined, value: undefined }`, and as a result I should test if `value` is `undefined` as well (but when that case arises and what to do in that case is unclear).

So I would say that the docs and the typing disagree, which is confusing for the users of `useAsync`: which one to believe?

To conclude with, for me this is a bug in the typing, so that's why I'm submitting a PR to fix it :relaxed:

---

_Side note: I have updated my code to have better types for `AsyncFn` as well (+ rebased on master)._

**Comment by s-thom** [comment]
I came across this issue as well, and ended up writing my own type definition that I just cast the result to.

```ts
type AsyncState<T> =
  | {
      loading: true;
    }
  | {
      loading: false;
      error: Error;
      value: undefined;
    }
  | {
      loading: false;
      error: undefined;
      value: T;
    };
```
The main difference between this and your solution is the loading state doesn't have `error` or `undefined`, which better matches the implementation. I've also set them to undefined in the opposite cases, which means the compiler doesn't yell at you when you do `state.value`.

**Comment by Rogdham** [comment]
> The main difference between this and your solution is the loading state doesn't have `error` or `undefined`, which better matches the implementation.

The reason there is, is so that we could do type destructuring:
```ts
const { loading, error, value } = useAsync(…);


```

> I've also set them explicitly to undefined, because `?: undefined` looks weird to me

This is to allow the following: `{ loading: false, value: 42 }` without specifying `error: undefined` for example.

**Comment by s-thom** [comment]
Huh, I hadn't considered those. I generally avoid destructuring (just personal taste, but I have been using t a bit more lately), but the `?:` use case is just something I hadn't considered.

**Comment by xobotyi** [comment]
IMO there should be one more state falg like `loaded` or loading should be renamed to `loaded` and has `boolean | null` type, where:
* `null` - loading not started
* `false` - loading strated but not done
* `true` - loading done

Because current state signature has ambiguous `loaded: false` state meaning both, not started and done states;

**Comment by Rogdham** [comment]
@streamich any thought on @xobotyi 's comment?

If I understand correctly the proposition: returned values of `useAsyncFn`:
 * initial value (assuming no `initialState` argument`):
```ts
{
  loaded: false,
  loading: false,
}
```
 * once the callback has been called:
```ts
{
  loaded: false,
  loading: true,
}
```
 * if the callback resolves:
```ts
{
  loaded: true,
  loading: false,
  value: 42,
}
```

 * if the callback rejects:
```ts
{
  loaded: true,
  loading: false,
  error: new Error('Something went wrong'),
}
```

**Comment by wardoost** [comment]
@xobotyi, @Rogdham Only adding a boolean for `loaded` is not enough, a promise can also throw `undefined`. We could add an `errored` key to fix this, which makes the hook's API bigger than it needs to be. It's what is proposed in https://github.com/streamich/react-use/issues/88#issuecomment-458964715 but with different a different signature.

A status key which indicates all these different states could also work, see https://github.com/streamich/react-use/issues/88#issuecomment-458979326. This would be a breaking change but seems like the most complete solution to me.

Feel free to open a PR for this or continue the conversation in https://github.com/streamich/react-use/issues/88 as this Pr is about type definitions.

File: src/useAsyncFn.ts [hunk 1]
```diff
@@ -18,17 +18,33 @@ export type AsyncState<T> =
...
 
-export type AsyncFn<Result = any, Args extends any[] = any[]> = [
-  AsyncState<Result>,
+export type AsyncFn<Result = any, Args extends any[] = any[], InitialState = Result> = [
+  AsyncState<Result | InitialState>,
   (...args: Args | []) => Promise<Result>
 ];
 
-export default function useAsyncFn<Result = any, Args extends any[] = any[]>(
+// initialState with value: returns AsyncFn<Result, Args>
```
Comment (wardoost) [inline src/useAsyncFn.ts]: Forgot to remove first 2 `useAsyncFn` function definitions?

Comment (wardoost) [inline src/useAsyncFn.ts]: Forgot to remove first 2 useAsyncFn function definitions?

File: src/useAsyncFn.ts [hunk 2]
```diff
@@ -18,17 +18,33 @@ export type AsyncState<T> =
       value: T;
```
Comment (wardoost) [inline src/useAsyncFn.ts]: Couldn't we just do `value: T | undefined;` on this line so we don't need to define an `InitialState` for the `AsyncFn` type and don't need to defined 2 version of `AsyncState` for the `initialState` parameter? A Promise can resolve `undefined` anyways 🤷‍♂️

**Comment by xobotyi** [comment]
@wardoost same as throw undefined it can fulfill with undefined %)
But its still be a 'loaded:true' + 'loading:false', this one is non-breaking change that solves the problem, despite the moment that I'd done as in comment you linked if I develop that hook.

**Comment by wardoost** [comment]
@xobotyi So we would have to add both a `loaded` and `errored` key to be able to determine if a promise throwed or resolved an `undefined` value as pointed out before.

It's easy to say "I would have done if I did it" especially considering the original was written back in October 2018 when hooks wasn't even released in the stable version of React. It's always best to be careful about introducing breaking changes but I think have a single status key should be up for consideration because if keeps the hook's API more clear despite being a breaking change. We could also add a `status` or `state` key to the state with `'idle' | 'pending' | 'rejected' | 'fulfilled'` as possible values and keep the loading key so it's not a breaking change, add a dev warning when the `loading` key is used and introduce the breaking change at a later change to give devs time to update existing code.

I think we best continue the conversation on this in #88 as this PR is about type definitions.

File: src/useAsyncFn.ts (hunk 1 above)
Comment (wardoost) [reply src/useAsyncFn.ts]: Didn't see this was function overloading, not using TS enough in daily work 😬

File: src/useAsyncFn.ts (hunk 2 above)
Comment (Rogdham) [reply src/useAsyncFn.ts]: I agree that Promise can resolve to `undefined`, but in that case `undefined` will be _inside_ `T`.

If we allow `value: T | undefined`, then developers will need to check against `undefined`, i.e. an problem similar to what this issue is about:

**Comment by Rogdham** [comment]
> Only adding a boolean for loaded is not enough, a promise can also throw undefined.

I did not considered that for the following reasons:
 1. The existing type already uses `Error` for the `error` key (which will not work if a Promise rejects with a number for example);
 2. I use the [`no-throw-literal` ESLint rule](https://eslint.org/docs/rules/no-throw-literal) which checks that Promises rejects `Error`s: _”It is considered good practice to only throw the Error object itself or an object using the Error object as base objects for user-defined exceptions.”_ I'm not sure if this is a well-established practice, or if it's ok to assume that in any way.

What do you think @wardoost? Especially concerning the point 1 above, i.e. existing type already asserts that Promise reject `Error`s.

**Comment by streamich** [comment]
__TLDR;__ maybe the types could be something like this:

```ts
export type AsyncState<T, E = Error> =
  | {
      state: 'pending';
      loading: true;
      error?: unknown;
      value?: unknown;
    }
  | {
      state: 'rejected';
      loading: false;
      error: E;
      value?: unknown;
    }
  | {
      state: 'fulfilled';
      loading: false;
      error?: unknown;
      value: T;
    };

export type AsyncStateFn<T, E = Error> =
  | {
      state: 'idle';
      loading?: unknown;
      error?: unknown;
      value?: unknown;
    }
  | AsyncState<T, E>;
```

---

1. The initial API for this hook was probably inspired by Apollo Client, [which returns `{ loading, error, data }`](https://www.apollographql.com/docs/react/data/queries/#executing-a-query), this hook swapped `data` for `value`, [which led to `{ loading, error, value }` tuple](https://github.com/streamich/react-use/commit/258d6964388c062b6bc0ee8db39d6d29eb072901#diff-8e6e0808e6745d8058a2f9cbd476172dR4-R6), which I believe still should work fine for the `useAsync` hook (possibly with some tweaks). It is the `useAsyncFn` hook that is more complex.

2. The original `useAsync` hook (and still now) executes the asynchronous function "immediately", which means that `{ loading, error }` should be enough to capture the state of the asynchronous function. Eventually, `useAsyncFn` hook was added, which does not execute the asynchronous function immediately, but instead user can choose when to call it. This adds extra state, which currently, I believe, is not fully captured by the `{ loading, error }` tuple. Hence, for example @xobotyi, has proposed to add an extra `{ loaded }` state.

3. Currently, `useAsync` uses `useAsyncFn` hook internally, I wonder if `useAsyncFn` should use `useAsync` instead, or maybe even the two hooks could be separate.

4. The error property does not necessarily need to be of type `Error`, maybe that's why in the original implementation it was `Error | any`. We should still allow for the possibility of `error` not being of type `Error`, maybe using `Error` as the default type, but allowing user to overwrite it, like `const useAsync = <E = Error>() => { ... }`.

5. Another case which current `useAsync` implementation does not really cover (as mentioned by @wardoost  and others) is that if `error` itself is of type `undefined`. Imagine the asynchronous function throws `undefined`, the state would be `{ loading: false, error: undefined, value: undefined }`. Did promise *resolve* or *reject*? One could, of course, check if the `undefined` values on `error` and `value` are own-properties or missing keys, but that is 🤯. So, even the state of `useAsync` hook today probably does not capture correctly all cases.

6. Probably `{ loading, error }` captures the state of the operation correctly in the Apollo Client, as their `error` is always of type `ApolloError`, but in `react-use` we need to deal with all possible cases. Hence maybe we could [add a more specific `state` property](https://github.com/streamich/react-use/issues/88#issuecomment-458979326), which would *precisely identify* the state; `error` and `value` would not be used for state identification&mdash;those would just hold data. Also, we could still keep `loading` (but deprecated), so it wold not be a breaking change (like suggested by @wardoost and possibly name it `status`, not `state`).

7. As mentioned by @Rogdham, we should keep all properties in all unioned interfaces that create the `AsyncState` type, so we can do destructuring `const { loading, value, error } = useAsync(...)`.

8. Maybe we could add `onFinally` effect [like suggested here](https://github.com/streamich/react-use/issues/88#issuecomment-466614687).

9. There is `useAsyncRetry` hook, but no `useAsyncFnRetry`; I wonder if we need one.
